File: kanonak-wire/go/wire.go
// Package wire is the Go port of the Kanonak wire kernel
// (kanonak.org/wire-form, wireFormatVersion "1").
//
// A minimal, allocation-conscious binary reader/writer for hot-path wire
// protocols. Generated protocol codecs call this kernel; it contains only
// what is invariant across ALL protocols — bounds-checked cursor
// reads/writes, big-endian integers, strict text validation, and a rich
// error taxonomy.
//
// Zero-copy contract: Bytes(n) and Rest() return subslices of the source
// buffer, never copies. Fail-loud contract: every failure is a *WireError
// stating what was expected, what was found, and where. Writer numeric
// parameters use exact-width types (byte/uint16/uint32) — the type is the
// range validation.
package wire

import (
	"fmt"
	"unicode/utf8"
)

// WireFormatVersion is the wire-form contract version this port implements.
const WireFormatVersion = "1"

// The error taxonomy of wireFormatVersion "1".
const (
	KindTruncated       = "Truncated"
	KindLengthOverrun   = "LengthOverrun"
	KindTrailingBytes   = "TrailingBytes"
	KindInvalidUtf8     = "InvalidUtf8"
	KindInvalidUuid     = "InvalidUuid"
	KindValueOutOfRange = "ValueOutOfRange"
	KindUnknownTag      = "UnknownTag"
)

// WireError is a wire kernel error: what was expected, what was found, and
// where. It works with errors.As.
type WireError struct {
	// Kind is one of the Kind* constants.
	Kind string
	// Offset is the absolute byte offset where the failing read started
	// (read-side errors). Meaningful only when HasOffset is true.
	Offset    int
	HasOffset bool
	message   string
}

func (e *WireError) Error() string {
	return e.message
}

func truncatedError(needed, remaining, offset int, context string) *WireError {
	return &WireError{
		Kind:      KindTruncated,
		Offset:    offset,
		HasOffset: true,
		message: fmt.Sprintf(
			"Truncated: %s needs %d byte(s) at offset %d, %d remain",
			context, needed, offset, remaining),
	}
}

func lengthOverrunError(declared, remaining, offset int, context string) *WireError {
	return &WireError{
		Kind:      KindLengthOverrun,
		Offset:    offset,
		HasOffset: true,
		message: fmt.Sprintf(
			"LengthOverrun: %s at offset %d declares %d byte(s), %d remain after the length field",
			context, offset, declared, remaining),
	}
}

func trailingBytesError(count, offset int) *WireError {
	return &WireError{
		Kind:      KindTrailingBytes,
		Offset:    offset,
		HasOffset: true,
		message: fmt.Sprintf(
			"TrailingBytes: expected end of buffer at offset %d, %d byte(s) remain",
			offset, count),
	}
}

func invalidUtf8ReadError(offset int, context string) *WireError {
	return &WireError{
		Kind:      KindInvalidUtf8,
		Offset:    offset,
		HasOffset: true,
		message:   fmt.Sprintf("InvalidUtf8: %s at offset %d", context, offset),
	}
}

func invalidUtf8WriteError(context string) *WireError {
	return &WireError{
		Kind:    KindInvalidUtf8,
		message: fmt.Sprintf("InvalidUtf8: %s", context),
	}
}

func invalidUuidError(context string) *WireError {
	return &WireError{
		Kind:    KindInvalidUuid,
		message: fmt.Sprintf("InvalidUuid: %s", context),
	}
}

func valueOutOfRangeError(value int, typeName string) *WireError {
	return &WireError{
		Kind:    KindValueOutOfRange,
		message: fmt.Sprintf("ValueOutOfRange: %d is not a valid %s", value, typeName),
	}
}

// NewUnknownTagError is the constructor for generated union dispatch on an
// unrecognized tag byte (not exercised by the kernel vectors).
func NewUnknownTagError(tag byte, context string) *WireError {
	return &WireError{
		Kind:    KindUnknownTag,
		message: fmt.Sprintf("UnknownTag: 0x%02x is not a known %s", tag, context),
	}
}

const hexLower = "0123456789abcdef"

// WireReader is a bounds-checked cursor over an immutable byte buffer.
// It never copies.
type WireReader struct {
	buf []byte
	pos int
}

// NewWireReader wraps buf without copying it.
func NewWireReader(buf []byte) *WireReader {
	return &WireReader{buf: buf}
}

func (r *WireReader) need(n int, context string) *WireError {
	remaining := len(r.buf) - r.pos
	if remaining < n {
		return truncatedError(n, remaining, r.pos, context)
	}
	return nil
}

// U8 reads one byte, 0..255.
func (r *WireReader) U8() (byte, error) {
	if err := r.need(1, "u8"); err != nil {
		return 0, err
	}
	v := r.buf[r.pos]
	r.pos++
	return v, nil
}

// U16BE reads two bytes, big-endian, 0..65535.
func (r *WireReader) U16BE() (uint16, error) {
	if err := r.need(2, "u16be"); err != nil {
		return 0, err
	}
	v := uint16(r.buf[r.pos])<<8 | uint16(r.buf[r.pos+1])
	r.pos += 2
	return v, nil
}

// U32BE reads four bytes, big-endian, 0..2^32-1.
func (r *WireReader) U32BE() (uint32, error) {
	if err := r.need(4, "u32be"); err != nil {
		return 0, err
	}
	p := r.pos
	v := uint32(r.buf[p])<<24 | uint32(r.buf[p+1])<<16 | uint32(r.buf[p+2])<<8 | uint32(r.buf[p+3])
	r.pos += 4
	return v, nil
}

// Bytes reads exactly n bytes as a zero-copy subslice of the source buffer.
func (r *WireReader) Bytes(n int) ([]byte, error) {
	if err := r.need(n, fmt.Sprintf("bytes(%d)", n)); err != nil {
		return nil, err
	}
	v := r.buf[r.pos : r.pos+n]
	r.pos += n
	return v, nil
}

// UUID reads 16 bytes as a lowercase hyphenated 8-4-4-4-12 UUID string.
// Any 16 bytes are legal — no version/variant validation.
func (r *WireReader) UUID() (string, error) {
	if err := r.need(16, "uuid"); err != nil {
		return "", err
	}
	b := r.buf[r.pos : r.pos+16]
	r.pos += 16
	out := make([]byte, 0, 36)
	for i, byt := range b {
		if i == 4 || i == 6 || i == 8 || i == 10 {
			out = append(out, '-')
		}
		out = append(out, hexLower[byt>>4], hexLower[byt&0x0f])
	}
	return string(out), nil
}

// UTF8 reads n bytes decoded as STRICT UTF-8 — invalid sequences, overlong
// encodings, and surrogate-range encodings are InvalidUtf8; never lossy,
// never U+FFFD. Bounds are checked before validity.
func (r *WireReader) UTF8(n int) (string, error) {
	start := r.pos
	view, err := r.Bytes(n)
	if err != nil {
		return "", err
	}
	if !utf8.Valid(view) {
		r.pos = start // the read did not take effect
		return "", invalidUtf8ReadError(start, fmt.Sprintf("utf8(%d)", n))
	}
	return string(view), nil
}

// LenPrefixedBytes16 reads a u16be length L, then exactly L bytes as a
// zero-copy subslice. L > remaining is LengthOverrun (NOT Truncated — the
// length field itself is suspect).
func (r *WireReader) LenPrefixedBytes16() ([]byte, error) {
	start := r.pos
	if err := r.need(2, "lenPrefixedBytes16"); err != nil {
		return nil, err
	}
	declared := int(r.buf[r.pos])<<8 | int(r.buf[r.pos+1])
	remainingAfterLength := len(r.buf) - r.pos - 2
	if declared > remainingAfterLength {
		return nil, lengthOverrunError(declared, remainingAfterLength, start, "lenPrefixedBytes16")
	}
	r.pos += 2
	v := r.buf[r.pos : r.pos+declared]
	r.pos += declared
	return v, nil
}

// Rest returns all remaining bytes (possibly empty) as a zero-copy subslice.
// It never errors and advances the cursor to the end.
func (r *WireReader) Rest() []byte {
	v := r.buf[r.pos:]
	r.pos = len(r.buf)
	return v
}

// Remaining is the count of unread bytes.
func (r *WireReader) Remaining() int {
	return len(r.buf) - r.pos
}

// ExpectEnd errors TrailingBytes if any bytes remain.
func (r *WireReader) ExpectEnd() error {
	count := len(r.buf) - r.pos
	if count > 0 {
		return trailingBytesError(count, r.pos)
	}
	return nil
}

// WireWriter is an append-only buffer builder. Numeric parameters use
// exact-width types — the type is the range validation.
type WireWriter struct {
	buf []byte
}

// NewWireWriter creates an empty writer.
func NewWireWriter() *WireWriter {
	return &WireWriter{}
}

// NewWireWriterWithCapacity preallocates so generated encoders can compute
// exact sizes and write once.
func NewWireWriterWithCapacity(capacity int) *WireWriter {
	return &WireWriter{buf: make([]byte, 0, capacity)}
}

// U8 appends one byte.
func (w *WireWriter) U8(v byte) *WireWriter {
	w.buf = append(w.buf, v)
	return w
}

// U16BE appends two bytes, big-endian.
func (w *WireWriter) U16BE(v uint16) *WireWriter {
	w.buf = append(w.buf, byte(v>>8), byte(v))
	return w
}

// U32BE appends four bytes, big-endian.
func (w *WireWriter) U32BE(v uint32) *WireWriter {
	w.buf = append(w.buf, byte(v>>24), byte(v>>16), byte(v>>8), byte(v))
	return w
}

// Bytes appends b verbatim.
func (w *WireWriter) Bytes(b []byte) *WireWriter {
	w.buf = append(w.buf, b...)
	return w
}

// UUID parses hyphenated 8-4-4-4-12 hex, case-insensitively, and appends the
// 16 bytes. Anything else (un-hyphenated, wrong length, non-hex) is
// InvalidUuid.
func (w *WireWriter) UUID(s string) error {
	b := []byte(s)
	if len(b) != 36 || b[8] != '-' || b[13] != '-' || b[18] != '-' || b[23] != '-' {
		return invalidUuidError(fmt.Sprintf("%q is not a hyphenated 8-4-4-4-12 UUID", s))
	}
	var out [16]byte
	oi := 0
	for i := 0; i < 36; {
		if b[i] == '-' {
			i++
			continue
		}
		hi, hiOK := hexVal(b[i])
		lo, loOK := hexVal(b[i+1])
		if !hiOK || !loOK {
			return invalidUuidError(fmt.Sprintf("%q is not a hyphenated 8-4-4-4-12 UUID", s))
		}
		out[oi] = hi<<4 | lo
		oi++
		i += 2
	}
	w.buf = append(w.buf, out[:]...)
	return nil
}

// UTF8 appends the UTF-8 encoding of s. Go strings can hold arbitrary bytes,
// so the never-lossy rule requires validation: an ill-formed string is
// InvalidUtf8 — never replacement characters.
func (w *WireWriter) UTF8(s string) error {
	if !utf8.ValidString(s) {
		return invalidUtf8WriteError("string is not well-formed UTF-8")
	}
	w.buf = append(w.buf, s...)
	return nil
}

// LenPrefixedBytes16 appends a u16be length, then the bytes. A length above
// 0xFFFF is ValueOutOfRange.
func (w *WireWriter) LenPrefixedBytes16(b []byte) error {
	if len(b) > 0xffff {
		return valueOutOfRangeError(len(b), "lenPrefixedBytes16 length")
	}
	w.U16BE(uint16(len(b)))
	w.Bytes(b)
	return nil
}

// ToBytes returns the written bytes, exact length (a copy — the builder
// stays reusable).
func (w *WireWriter) ToBytes() []byte {
	out := make([]byte, len(w.buf))
	copy(out, w.buf)
	return out
}

func hexVal(c byte) (byte, bool) {
	switch {
	case c >= '0' && c <= '9':
		return c - '0', true
	case c >= 'a' && c <= 'f':
		return c - 'a' + 10, true
	case c >= 'A' && c <= 'F':
		return c - 'A' + 10, true
	default:
		return 0, false
	}
}

File: kanonak-wire/swift/Sources/KanonakWire/Wire.swift
// `KanonakWire` — the Swift port of the Kanonak wire kernel
// (`kanonak.org/wire-form`, `wireFormatVersion "1"`).
//
// A minimal, allocation-conscious binary reader/writer for hot-path wire
// protocols. Generated protocol codecs call this kernel; it contains only what
// is invariant across ALL protocols — bounds-checked cursor reads/writes,
// big-endian integers, strict text validation, and a rich error taxonomy. It
// knows nothing about any particular protocol.
//
// Zero-copy contract: `bytes(_:)`, `rest()`, and the byte view behind `utf8(_:)`
// return `ArraySlice<UInt8>` views into the source buffer, never copies —
// Swift's natural byte-view type (the counterpart of Rust's slice, Go's
// subslice, and TypeScript's `Uint8Array` subarray).
//
// Fail-loud contract: every failure is a `WireError` stating what was expected,
// what was found, and where. No silent fallbacks — no optional-nil returns, no
// partial values, no lossy decodes.
//
// Two representability notes, which is why the corresponding shared vectors are
// SKIPPED (with a reported count) rather than passed by this port:
//
//   - Writer numeric parameters are exact-width (`UInt8`/`UInt16`/`UInt32`), so
//     the type IS the range validation and out-of-range/non-integer arguments
//     cannot be constructed. (`wide-numeric-params`, `dynamic-numeric` — the
//     same position Rust, Go, and C# take.)
//   - A Swift `String` is a sequence of `Unicode.Scalar`s, and `Unicode.Scalar`
//     cannot hold a surrogate — so the lone-surrogate encoder trap that JS,
//     Java, C#, and Python must guard against is unrepresentable here, and
//     `utf8(_:)` on the writer is infallible. (`utf16-strings`.)

public let wireFormatVersion = "1"

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

/// The error taxonomy of `wireFormatVersion "1"`. Raw values are the kind names
/// as they appear in the shared vectors.
public enum WireErrorKind: String {
    case truncated = "Truncated"
    case lengthOverrun = "LengthOverrun"
    case trailingBytes = "TrailingBytes"
    case invalidUtf8 = "InvalidUtf8"
    case invalidUuid = "InvalidUuid"
    case valueOutOfRange = "ValueOutOfRange"
    case unknownTag = "UnknownTag"
}

/// A wire kernel error: what was expected, what was found, and where.
public struct WireError: Error, CustomStringConvertible {
    public let kind: WireErrorKind
    /// Absolute byte offset where the failing read started (read-side errors).
    public let offset: Int?
    public let message: String

    public var description: String { message }

    public static func truncated(needed: Int, remaining: Int, offset: Int, context: String) -> WireError {
        WireError(
            kind: .truncated,
            offset: offset,
            message: "Truncated: \(context) needs \(needed) byte(s) at offset \(offset), \(remaining) remain"
        )
    }

    public static func lengthOverrun(declared: Int, remaining: Int, offset: Int, context: String) -> WireError {
        WireError(
            kind: .lengthOverrun,
            offset: offset,
            message: "LengthOverrun: \(context) at offset \(offset) declares \(declared) byte(s), "
                + "\(remaining) remain after the length field"
        )
    }

    public static func trailingBytes(count: Int, offset: Int) -> WireError {
        WireError(
            kind: .trailingBytes,
            offset: offset,
            message: "TrailingBytes: expected end of buffer at offset \(offset), \(count) byte(s) remain"
        )
    }

    public static func invalidUtf8(offset: Int, context: String) -> WireError {
        WireError(kind: .invalidUtf8, offset: offset, message: "InvalidUtf8: \(context) at offset \(offset)")
    }

    public static func invalidUuid(context: String) -> WireError {
        WireError(kind: .invalidUuid, offset: nil, message: "InvalidUuid: \(context)")
    }

    public static func valueOutOfRange(value: Any, typeName: String) -> WireError {
        WireError(
            kind: .valueOutOfRange,
            offset: nil,
            message: "ValueOutOfRange: \(value) is not a valid \(typeName)"
        )
    }

    /// Constructor for generated union dispatch on an unrecognized tag byte.
    public static func unknownTag(tag: UInt8, context: String) -> WireError {
        WireError(
            kind: .unknownTag,
            offset: nil,
            message: "UnknownTag: 0x\(format2Hex(tag)) is not a known \(context)"
        )
    }
}

private let hexLower: [Character] = Array("0123456789abcdef")

/// Two lowercase hex digits for a byte — no Foundation dependency.
private func format2Hex(_ byte: UInt8) -> String {
    String([hexLower[Int(byte >> 4)], hexLower[Int(byte & 0x0f)]])
}

private func hexValue(_ c: UInt8) -> UInt8? {
    switch c {
    case 0x30...0x39: return c - 0x30              // 0-9
    case 0x61...0x66: return c - 0x61 + 10         // a-f
    case 0x41...0x46: return c - 0x41 + 10         // A-F
    default: return nil
    }
}

// ---------------------------------------------------------------------------
// Strict UTF-8 validation
// ---------------------------------------------------------------------------

/// Is this byte sequence well-formed UTF-8 (RFC 3629)? Rejects overlong
/// encodings, the surrogate range (U+D800–U+DFFF), values above U+10FFFF, and
/// truncated sequences. Written out rather than delegated so the strictness is
/// auditable here and cannot vary with a platform's decoder: `String(decoding:
/// as: UTF8.self)` is LOSSY (it substitutes U+FFFD), which this contract forbids.
private func isWellFormedUTF8<C: Collection>(_ bytes: C) -> Bool where C.Element == UInt8 {
    var i = bytes.startIndex
    let end = bytes.endIndex

    func continuation(_ idx: C.Index, _ lo: UInt8, _ hi: UInt8) -> Bool {
        idx < end && bytes[idx] >= lo && bytes[idx] <= hi
    }

    while i < end {
        let b = bytes[i]
        let width: Int
        switch b {
        case 0x00...0x7f:
            width = 1
        case 0xc2...0xdf:
            let c1 = bytes.index(after: i)
            guard continuation(c1, 0x80, 0xbf) else { return false }
            width = 2
        case 0xe0:
            // Second byte A0..BF excludes the overlong 3-byte forms.
            let c1 = bytes.index(after: i)
            guard continuation(c1, 0xa0, 0xbf), continuation(bytes.index(after: c1), 0x80, 0xbf) else { return false }
            width = 3
        case 0xe1...0xec, 0xee...0xef:
            let c1 = bytes.index(after: i)
            guard continuation(c1, 0x80, 0xbf), continuation(bytes.index(after: c1), 0x80, 0xbf) else { return false }
            width = 3
        case 0xed:
            // Second byte 80..9F excludes the surrogate range U+D800–U+DFFF.
            let c1 = bytes.index(after: i)
            guard continuation(c1, 0x80, 0x9f), continuation(bytes.index(after: c1), 0x80, 0xbf) else { return false }
            width = 3
        case 0xf0:
            // Second byte 90..BF excludes the overlong 4-byte forms.
            let c1 = bytes.index(after: i)
            let c2 = bytes.index(after: c1)
            guard continuation(c1, 0x90, 0xbf), continuation(c2, 0x80, 0xbf),
                  continuation(bytes.index(after: c2), 0x80, 0xbf) else { return false }
            width = 4
        case 0xf1...0xf3:
            let c1 = bytes.index(after: i)
            let c2 = bytes.index(after: c1)
            guard continuation(c1, 0x80, 0xbf), continuation(c2, 0x80, 0xbf),
                  continuation(bytes.index(after: c2), 0x80, 0xbf) else { return false }
            width = 4
        case 0xf4:
            // Second byte 80..8F caps the range at U+10FFFF.
            let c1 = bytes.index(after: i)
            let c2 = bytes.index(after: c1)
            guard continuation(c1, 0x80, 0x8f), continuation(c2, 0x80, 0xbf),
                  continuation(bytes.index(after: c2), 0x80, 0xbf) else { return false }
            width = 4
        default:
            // 0x80–0xC1 (lone continuation / overlong 2-byte) and 0xF5–0xFF.
            return false
        }
        i = bytes.index(i, offsetBy: width)
    }
    return true
}

// ---------------------------------------------------------------------------
// Reader
// ---------------------------------------------------------------------------

/// A bounds-checked cursor over an immutable byte buffer. Never copies.
public struct WireReader {
    private let buf: [UInt8]
    private var pos: Int

    public init(_ buf: [UInt8]) {
        self.buf = buf
        self.pos = 0
    }

    /// Count of unread bytes.
    public var remaining: Int { buf.count - pos }

    private func need(_ n: Int, _ context: String) throws {
        let left = remaining
        if left < n {
            throw WireError.truncated(needed: n, remaining: left, offset: pos, context: context)
        }
    }

    public mutating func u8() throws -> UInt8 {
        try need(1, "u8")
        let v = buf[pos]
        pos += 1
        return v
    }

    public mutating func u16BE() throws -> UInt16 {
        try need(2, "u16be")
        let v = UInt16(buf[pos]) << 8 | UInt16(buf[pos + 1])
        pos += 2
        return v
    }

    public mutating func u32BE() throws -> UInt32 {
        try need(4, "u32be")
        let v = UInt32(buf[pos]) << 24 | UInt32(buf[pos + 1]) << 16
            | UInt32(buf[pos + 2]) << 8 | UInt32(buf[pos + 3])
        pos += 4
        return v
    }

    /// Exactly n bytes as a zero-copy view of the source buffer.
    public mutating func bytes(_ n: Int) throws -> ArraySlice<UInt8> {
        try need(n, "bytes(\(n))")
        let v = buf[pos..<(pos + n)]
        pos += n
        return v
    }

    /// 16 bytes as a lowercase hyphenated UUID string. NO version/variant
    /// validation — any 16 bytes are legal.
    public mutating func uuid() throws -> String {
        try need(16, "uuid")
        var s = ""
        s.reserveCapacity(36)
        for i in 0..<16 {
            if i == 4 || i == 6 || i == 8 || i == 10 { s.append("-") }
            let byte = buf[pos + i]
            s.append(hexLower[Int(byte >> 4)])
            s.append(hexLower[Int(byte & 0x0f)])
        }
        pos += 16
        return s
    }

    /// n bytes decoded as STRICT UTF-8. Bounds are checked before validity, and
    /// the kernel never peeks past the requested extent — a multibyte character
    /// split by `n` is `InvalidUtf8`, not a longer read.
    public mutating func utf8(_ n: Int) throws -> String {
        let start = pos
        let view = try bytes(n)
        guard isWellFormedUTF8(view) else {
            pos = start  // the read did not take effect
            throw WireError.invalidUtf8(offset: start, context: "utf8(\(n))")
        }
        return String(decoding: view, as: UTF8.self)
    }

    /// u16be length L, then exactly L bytes (zero-copy view). L beyond the
    /// remaining bytes is `LengthOverrun`, not `Truncated` — the length field
    /// itself is what is suspect.
    public mutating func lenPrefixedBytes16() throws -> ArraySlice<UInt8> {
        let start = pos
        try need(2, "lenPrefixedBytes16")
        let declared = Int(UInt16(buf[pos]) << 8 | UInt16(buf[pos + 1]))
        let remainingAfterLength = buf.count - pos - 2
        if declared > remainingAfterLength {
            throw WireError.lengthOverrun(
                declared: declared,
                remaining: remainingAfterLength,
                offset: start,
                context: "lenPrefixedBytes16"
            )
        }
        pos += 2
        let v = buf[pos..<(pos + declared)]
        pos += declared
        return v
    }

    /// All remaining bytes (possibly empty), zero-copy. Never errors.
    public mutating func rest() -> ArraySlice<UInt8> {
        let v = buf[pos...]
        pos = buf.count
        return v
    }

    public func expectEnd() throws {
        let count = remaining
        if count > 0 {
            throw WireError.trailingBytes(count: count, offset: pos)
        }
    }
}

// ---------------------------------------------------------------------------
// Writer
// ---------------------------------------------------------------------------

/// An append-only buffer builder. Numeric parameters use exact-width types —
/// the type is the range validation.
public struct WireWriter {
    private var buf: [UInt8]

    public init() { buf = [] }

    /// Preallocate, so a generated encoder that knows its exact size writes once.
    public init(capacity: Int) {
        buf = []
        buf.reserveCapacity(capacity)
    }

    public mutating func u8(_ value: UInt8) {
        buf.append(value)
    }

    public mutating func u16BE(_ value: UInt16) {
        buf.append(UInt8(truncatingIfNeeded: value >> 8))
        buf.append(UInt8(truncatingIfNeeded: value))
    }

    public mutating func u32BE(_ value: UInt32) {
        buf.append(UInt8(truncatingIfNeeded: value >> 24))
        buf.append(UInt8(truncatingIfNeeded: value >> 16))
        buf.append(UInt8(truncatingIfNeeded: value >> 8))
        buf.append(UInt8(truncatingIfNeeded: value))
    }

    public mutating func bytes<C: Collection>(_ b: C) where C.Element == UInt8 {
        buf.append(contentsOf: b)
    }

    /// Hyphenated 8-4-4-4-12 hex, case-insensitive input; emits the 16 bytes.
    /// The full shape is validated BEFORE any parsing — hyphens at exactly
    /// positions 8/13/18/23 and hex digits everywhere else — so a stray hyphen
    /// is `InvalidUuid`, never a short or shifted parse.
    public mutating func uuid(_ s: String) throws {
        let b = Array(s.utf8)
        let shapeOK = b.count == 36 && b.enumerated().allSatisfy { i, c in
            switch i {
            case 8, 13, 18, 23: return c == 0x2d  // '-'
            default: return hexValue(c) != nil
            }
        }
        guard shapeOK else {
            throw WireError.invalidUuid(context: "\"\(s)\" is not a hyphenated 8-4-4-4-12 UUID")
        }
        var out = [UInt8]()
        out.reserveCapacity(16)
        var i = 0
        while i < 36 {
            if b[i] == 0x2d { i += 1; continue }
            out.append(hexValue(b[i])! << 4 | hexValue(b[i + 1])!)
            i += 2
        }
        buf.append(contentsOf: out)
    }

    /// UTF-8 encode. A Swift `String` is well-formed by construction (its
    /// scalars cannot be surrogates) — infallible, as in Rust.
    public mutating func utf8(_ s: String) {
        buf.append(contentsOf: s.utf8)
    }

    /// u16be length, then the bytes. A length above 0xFFFF is `ValueOutOfRange`.
    public mutating func lenPrefixedBytes16<C: Collection>(_ b: C) throws where C.Element == UInt8 {
        let count = b.count
        guard count <= 0xffff else {
            throw WireError.valueOutOfRange(value: count, typeName: "lenPrefixedBytes16 length")
        }
        u16BE(UInt16(count))
        bytes(b)
    }

    /// The written bytes.
    public func toBytes() -> [UInt8] { buf }
}
